### ISSUE TITLE
Extend Sparse spatial data structures

### ISSUE THREAD

**Issue by stamnug** [taichi-dev/taichi#4310]
The document said that "When deactivation happens, the Taichi runtime automatically recycles and zero-fills memory of the deactivated containers." 

I hope supplied a temporary deactivate for save container's data and a long time deactivate for save container's data to the disk and automatically load them when container activated.

Or supplied a method be triggered for custom function when ti.activate and ti.deactivate be called.

**Comment by bobcao3** [comment]
the custom function, do you mean by a python function or a taichi function?

**Comment by k-ye** [comment]
Note that deactivation/recycling happens only after the kernel execution, therefore you might be able to just save the (deactivated) to disk after the kernel execution?

**Comment by stamnug** [comment]
> the custom function, do you mean by a python function or a taichi function?

currently, i just want to get the data while needed and deactivate container for save memory but save data for later while not needed.  i not have enough knowledge to decide how to implement more feasible, efficiency, flexible.

**Comment by k-ye** [comment]
If you just want to save memory and offload some data onto the disk, calling `to_numpy()`, then saving the NP array to disk is the easiest solution.

**Comment by stamnug** [comment]
> If you just want to save memory and offload some data onto the disk, calling `to_numpy()`, then saving the NP array to disk is the easiest solution.

yes, i just want to save memory and offload some data onto the disk, but to_numpy is inefficient and complex. maybe it is easy enough for a limited sized 'field'. but if the Sparse Spatial Data Structures as a more universal data structure, for example UE5 world partition system, i don't know whether this example is suitable , i naively think the SSDS be able to do it. i place a huge model in a field, using to_numpy will is unimaginable.

**Comment by k-ye** [comment]
> i place a huge model in a field, using to_numpy will is unimaginable.

Hmm, that's right. On top of my head, we can dump a sparse field into a more compact format, such as COO or CSR.

**Comment by stamnug** [comment]
Hmm, Maybe shound support variable length member in ti.Struct.field first. otherwise it is a problem that how place a model in a field. 😂

**Comment by k-ye** [comment]
@stamnug could you actually post your email reply here :-)? (It's fine if you'd prefer 中文)

**Comment by stamnug** [comment]
@k-ye  987004590@qq.com  🙃

**Comment by k-ye** [comment]
> 没错的是field里放个高模不管是从内存往显存里转，还是显存往内存里转开销估计都很高，最好就是最初构建一次，然后longtime deactive自动回收时显存直接offload到硬盘，active时直接从硬盘load到显存，估计还要考虑数据压缩啥的就交给大佬们琢磨了:p

是的。这里主要的区别是，deactivate 和"自动落到硬盘"是两个动作。因此更可取的方法是：deactivate 后从 field 转出一个更加紧凑的、可序列化的数据，然后把这个数据存到硬盘上 :-)

**Comment by stamnug** [comment]
@k-ye 
额..我把这内容删了..因为我最初直觉是field在显存里, deactivated是节省显存, 后来我又感觉field在内存里, 显卡通过DMA访问的内存, deactivated是节省内存. 压根就没有什么显存和内存传递数据问题.  只有field序列化存储的问题. 显卡的架构是我的知识盲区, 所以, 你说的对. 只需要将deactive pointer snode内容序列化出来即可.

**Comment by stamnug** [comment]
@k-ye 
不过, 序列化出来的数据数量可能是个问题, 一个pointer序列化出来一个数据, 交由程序员管理恐怕是个大累赘, 如果没有足够的动机, 没必要把这个东西暴露给程序员.

**Comment by k-ye** [comment]
https://github.com/taichi-dev/taichi/issues/4310#issuecomment-1046481009

没必要删除，我们目的就是通过 github issue/discussion 不断探讨一些方案的  :-) 

> 不过, 序列化出来的数据数量可能是个问题, 一个pointer序列化出来一个数据, 交由程序员管理恐怕是个大累赘, 如果没有足够的动机, 没必要把这个东西暴露给程序员.

是的嗯。你可以把这个序列化出来的东西理解为一个 opaque data structure，唯一的目的就是：
* 可以在这个 opaque data structure 和 sparse field 之间转换
* 可以把这个 opaque ds 存到硬盘/从磁盘读出来

**Comment by stamnug** [comment]
> 是的嗯。你可以把这个序列化出来的东西理解为一个 opaque data structure，提供的功能就是：
> 
>     * 可以在这个 opaque data structure 和 sparse field 之间转换
> 
>     * 可以把这个 opaque ds 存到硬盘/从磁盘读出来

No problem.
